### PR TITLE
Simplify generics on Mapper.Builder

### PR DESCRIPTION
**Pull request by romseygeek**
Mapper.Builder currently has some complex generics on it to allow fluent builder
construction.  However, the second parameter, a return type from the `build()` method,
is unnecessary, as we can use covariant return types.  This commit removes this second
generic parameter.

### CODE REVIEW

**Comment by romseygeek** [comment]
I think we can probably remove the first generic parameter as well, if we're willing to give up on fluent builders - it adds a lot of complexity and visual noise for not much return, IMO.

File: server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java [hunk 1]
```diff
@@ -45,10 +45,12 @@
     }
 
     @SuppressWarnings("rawtypes")
-    public abstract static class Builder<T extends Builder, Y extends MetadataFieldMapper> extends FieldMapper.Builder<T, Y> {
+    public abstract static class Builder<T extends Builder> extends FieldMapper.Builder<T> {
         public Builder(String name, MappedFieldType fieldType, MappedFieldType defaultFieldType) {
             super(name, fieldType, defaultFieldType);
         }
+
+        public abstract MetadataFieldMapper build(BuilderContext context);
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java:53]: just out of curiosity: this seems to be one of the few reasons where the generic `build` return type was needed?

**Review by cbuescher — COMMENTED** [review]
Great simplification, I was looking for the spots where we actually needed that second parameters and could only find one where I'm even not sure (left a short question just for my understanding). LGTM from my side.
I have one question regarding backport to 7.9 though: are we allowing users to define and add custom field types to Elasticsearch, e.g. via Plugins? In this case this might be a breaking change I think. Not sure about this though.

File: server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java (hunk 1 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/MetadataFieldMapper.java:53]: There are a few places in tests as well that expect build() to return a specific type, but nothing within the server code itself.

**Comment by romseygeek** [comment]
> are we allowing users to define and add custom field types to Elasticsearch, e.g. via Plugins?

Yes, this would be a breaking change, but I think we allow these in point releases? That's why plugins have to be tied to a specific release.

I've actually tried removing the first generic parameter as well, and that adds very little noise to the PR but makes things a whole lot simpler still - should I include that here, or open up a followup do you think?

**Comment by cbuescher** [comment]
> I think we allow these in point releases? That's why plugins have to be tied to a specific release.

Tbh I'm not sure, just wanted to bring this up since it came to mind. Maybe needs another opinion.

> I've actually tried removing the first generic parameter as well

I wouldn't mind if its in a separate commit. Depending on the diff it might make sense to open another PR though, also maybe for folks who e.g. later want to track the history of an API change.

**Comment by romseygeek** [comment]
It's mostly changes on the same lines as this current one - I'll push the new changes here.

**Comment by romseygeek** [comment]
Have reverted the complete generics removal, and I'll open a separate issue for that.

**Review by cbuescher — APPROVED** [review]
Thanks, LGTM to the first part

**Pull request merged**